### PR TITLE
class orderings

### PR DESCRIPTION
**Pull request by ieaves**
This PR introduces the notion of ordering among complexity classes through a `__gt__` and `__lt__` method on the base complexity class. Each concrete implementation of a ComplexityClass is assumed to have an `order` attribute specifying their relative order among other implementations.

### CODE REVIEW

**Comment by pberkes** [comment]
Hi @ieaves thank you for your PR. What is the order based on? Is there a formal way of defining an order?

**Comment by ieaves** [comment]
Hey @pberkes, practically speaking the order is intended to answer "which runs faster," stated [formally:](https://plato.stanford.edu/entries/computational-complexity/#BasCon)


> The efficiency of two machines is compared according to the order of growth of their time and space complexities. In particular, given a function f:ℕ→ℕ we define its order of growth to be O(f(n))={g(n):∃c∃n0∀n≥n0(g(n)<c⋅f(n))} – i.e. the set of all functions which are asymptotically bounded by f(n), ignoring scalar factors. For instance, for all fixed k,c1,c2 ∈ℕ the following functions are all in O(n2): the constant k-function, log(n),n,c1⋅n2+c2. However 11000n3∉O(n2). A machine M1 is said to have lower time complexity (or to run faster than) another machine M2 if tM1(n)∈O(tM2(n)), but not conversely.

In other words, given two complexity class f1 and f2, f1 > f2 if there exists a k such that f1(n) > f2(n) for all values of n > k.

**Comment by pberkes** [comment]
If I understand correctly, there is an order but the values you are using in the implementations are arbitrary. Is that correct?

If that's the case, what do you think of spacing them by 10? as in 30, 40, 50, ... like the good old line numbering in BASIC? In that way, if we introduce a new complexity class that is somewhere in between two current classes we don't have to modify all the constants

**Comment by pberkes** [comment]
Otherwise I'm I think all that is needed before merging is a good test, defining 2-3 dummy complexity classes and testing all operations <, >, >=, <= .

**Comment by ieaves** [comment]
> If I understand correctly, there is an order but the values you are using in the implementations are arbitrary. Is that correct?

That's correct, if each ComplexityClass was guaranteed to have an associated symbolic representation we could analytically compute orderings but this seemed like an easier solution when the potential classes was known in advance.

> If that's the case, what do you think of spacing them by 10? as in 30, 40, 50, ... like the good old line numbering in BASIC?

That works just fine! My original thought was for the ordering to correspond to the default set of complexity classes in the package reserving decimal values for user customized classes but at the end of the day it's still arbitrary :).

> Otherwise I'm I think all that is needed before merging is a good test, defining 2-3 dummy complexity classes and testing all operations <, >, >=, <= , and == .

A quick note, I specifically didn't modify `==` because doing so requires me to also implement hash on the base `ComplexityClass`. Otherwise, I can add a few tests for the purpose!

**Comment by ieaves** [comment]
The quadratic test case was consistently failing on local runs so I also increased the test length to 500 elements which passed.

**Comment by ieaves** [comment]
@pberkes If this PR meets requirements to merge would you be so kind as to create a new pypi release? I would like to include this as a testing requirement on a [different project](https://github.com/dylan-profiler/visions) I'm involved with. If you need anything else from me to make that happen please don't hesitate to ask.

**Comment by pberkes** [comment]
My comment on `==` is that you use it in `__ge__` and `__le__` but it's not defined, so if you try and compare two classes with the same complexity you would get `False`.

File: big_o/test/test_complexityclass_comparisons.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        self.assertEqual(SecondComplexity < FirstComplexity, False)
+
+    def test_g(self):
+        self.assertEqual(FirstComplexity > SecondComplexity, False)
+        self.assertEqual(FirstComplexity > FirstComplexity, False)
+        self.assertEqual(SecondComplexity > FirstComplexity, True)
+
+    def test_eq(self):
+        self.assertEqual(FirstComplexity == SecondComplexity, False)
+        self.assertEqual(FirstComplexity == FirstComplexity, True)
```
Comment (pberkes) [inline big_o/test/test_complexityclass_comparisons.py:44]: To be clear again about the `<=`, `>=`, and `==`, I'm attaching a modified version of the tests to show how I expect the classes to behave. As you say, we can discuss whether the `==` test should pass, but I think the `>=` and `<=` tests should definitively pass (they are failing now)

Comment (pberkes) [reply big_o/test/test_complexityclass_comparisons.py:44]:
```
import unittest

from big_o.complexities import ComplexityClass


class FirstComplexityClass(ComplexityClass):
    order = 1


class AltFirstComplexityClass(ComplexityClass):
    order = 1


class SecondComplexityClass(ComplexityClass):
    order = 2


class TestComplexities(unittest.TestCase):

    def setUp(self):
        self.first_complexity = FirstComplexityClass()
        self.alt_first_complexity = AltFirstComplexityClass()
        self.second_complexity = SecondComplexityClass()

    def test_ge(self):
        self.assertFalse(self.first_complexity >= self.second_complexity)
        self.assertTrue(self.second_complexity >= self.first_complexity)
        self.assertTrue(self.first_complexity >= self.first_complexity)
        self.assertTrue(self.alt_first_complexity >= self.first_complexity)
        self.assertTrue(self.first_complexity >= self.alt_first_complexity)

    def test_le(self):
        self.assertTrue(self.first_complexity <= self.second_complexity)
        self.assertFalse(self.second_complexity <= self.first_complexity)
        self.assertTrue(self.first_complexity <= self.first_complexity)
        self.assertTrue(self.alt_first_complexity <= self.first_complexity)
        self.assertTrue(self.first_complexity <= self.alt_first_complexity)

    def test_l(self):
        self.assertTrue(self.first_complexity < self.second_complexity)
        self.assertFalse(self.second_complexity < self.first_complexity)
        self.assertFalse(self.first_complexity < self.alt_first_complexity)

    def test_g(self):
        self.assertFalse(self.first_complexity > self.second_complexity)
        self.assertTrue(self.second_complexity > self.first_complexity)
        self.assertFalse(self.first_complexity > self.alt_first_complexity)

    def test_eq(self):
        self.assertFalse(self.first_complexity == self.second_complexity)
        self.assertTrue(self.first_complexity == self.first_complexity)
        self.assertTrue(self.first_complexity == self.alt_first_complexity)
```

**Comment by ieaves** [comment]
Equality is defined by default, it defaults to hash comparison!

**Comment by ieaves** [comment]
What do you mean the tests are failing now? They all work for me on 3.7.
can you describe your environment?

File: big_o/test/test_complexityclass_comparisons.py (hunk 1 above)
Comment (ieaves) [reply big_o/test/test_complexityclass_comparisons.py]: Tests now pass.

**Comment by ieaves** [comment]
While I'm not sure what semantic meaning the AltFirstComplexityClass would have, the tests now behave as you prefer.

File: big_o/test/test_complexityclass_comparisons.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class SecondComplexityClass(ComplexityClass):
+    order = 2
+
+
+class TestComplexities(unittest.TestCase):
+
+    def setUp(self):
+        print('here')
```
Comment (pberkes) [inline big_o/test/test_complexityclass_comparisons.py:22]: Please remove

File: big_o/test/test_complexityclass_comparisons.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import unittest
+import numpy as np
```
Comment (pberkes) [inline big_o/test/test_complexityclass_comparisons.py:2]: Unused import

**Comment by pberkes** [comment]
LGTM

**Comment by ieaves** [comment]
Thank you @pberkes !

**Pull request merged**